bin/generate-layer.sh

```diff
@@ -0,0 +1,102 @@
+#!/usr/bin/env bash
+
+set -e
+
+# set defaults
+PYTHON_VERSION="python3.7"
+
+
+get_target_dirs() {
+  # create an array of all unique directories containing requirements.txt files
+  # within this project
+  PROJECT_DIR=$(cd "$( dirname "${BASH_SOURCE[0]}" )"/.. >/dev/null 2>&1 && pwd )
+  mapfile -t directories < <(find "$PROJECT_DIR" -name 'requirements.txt' -exec dirname {} \; | sort -u | grep -v "\.terraform\b\|\.terragrunt-cache\b")
+}
+
+build_image() {
+  echo "$(tput setaf 2)=> building image from $DOCKERFILE$(tput sgr0)"
+  docker build -t "$IMAGE_NAME" -f "$DOCKERFILE" .
+}
+
+
+docker_image_check() {
+  # validate if the docker image is present on the host machine
+  # if not found, create it
+  IMAGE_NAME="${IMAGE_NAME:=ldapamaint/amazonlinux}"
+  DOCKERFILE="Dockerfile.layers"
+
+  IMAGE_ID=$(docker inspect --type=image -f '{{.Id}}' "$IMAGE_NAME" 2> /dev/null || true)
+  if [ -z "$IMAGE_ID" ]; then
+    echo "$(tput setaf 4) > image $IMAGE_NAME doesn't exist$(tput sgr0)"
+    build_image
+  else
+    echo "$(tput setaf 4) > image $IMAGE_NAME already exists$(tput sgr0)"
+  fi
+}
+
+get_docker_bindmount_dirs() {
+  # create bindmount flags for each directory that requires a layer
+  get_target_dirs
+
+  CODE_DIR="/home/lambda_layer"
+  DOCKER_BINDMOUNTS=()
+
+  for dir in "${directories[@]}"
+  do
+    CONTAINER_DIR="$CODE_DIR/$(basename "${dir}")"
+    DOCKER_BINDMOUNTS+=(-v "${dir}:$CONTAINER_DIR")
+  done
+}
+
+docker_run() {
+  # run docker
+  docker_image_check
+  get_docker_bindmount_dirs
+  DOCKER_BINDMOUNTS+=(-v "$PROJECT_DIR/bin:$CODE_DIR/bin")
+
+  docker run --rm -ti \
+    "${DOCKER_BINDMOUNTS[@]}" \
+    -w "$CODE_DIR" \
+    --cap-add=SYS_PTRACE \
+    "$IMAGE_NAME" bash -c './bin/generate-layer.sh -c'
+}
+
+create_layers() {
+  get_target_dirs
+
+  command -v zip > /dev/null 2>&1 || (echo "zip not available"; exit 1)
+  command -v pip3 > /dev/null 2>&1 || (echo "pip3 not available"; exit 1)
+
+  for dir in "${directories[@]}"
+  do
+    cd "${dir}" || ( echo "Unable to navigate to ${dir}"; exit 1)
+    lambda_pacakge="/$(pwd)/lambda-package"
+    layer_path="$lambda_pacakge/python/lib/$PYTHON_VERSION/site-packages/"
+    mkdir -p "$layer_path" >/dev/null 2>&1 || ( echo "Unable to create $layer_path"; exit 1)
+    pip3 install -r requirements.txt -t "$layer_path" >/dev/null 2>&1 || (echo "Encountered error installing python dependency"; exit 1)
+    pushd lambda-package/ >/dev/null 2>&1 || ( echo "Unable to navigate to lambda-package/"; exit 1)
+    zip -r ../lambda_layer_payload.zip python/* -x "setuptools*/*" "pkg_resources/*" "easy_install*" >/dev/null 2>&1
+    popd >/dev/null 2>&1 || ( echo "Unable to return to source directory"; exit 1)
+    rm -rf "$lambda_pacakge"
+    echo "Generated layer for: $(basename "${dir}")"
+  done
+}
+
+while getopts :rc opt
+do
+    case "${opt}" in
+        r)
```

Comment:
so I guess this solidifies the over-complication I have going on here with this script.. but the idea was that I didn't want two separate files for managing layers. The short explanation is that the `-r` option calls docker and then runs this script again with the `-c` option within the container itself. 

full order of operations for the `-r` option (run docker):
1. check if the `ldapamaint/amazonlinux` image exists in the local docker registry, else create it
2. one dir up from the script execution path (`project-name/bin`), grab a list of dirs that contain a `requirements.txt` file
3. run docker with bindmounts for each of the dirs collected in step 2 and call `generate-layer.sh` with the `-c` option

`-c` option (create layers):
1. one dir up from the script execution path, grab a list of dirs that contain a `requirements.txt` file
2. for each dir in the resulting list run `pip install` and direct the contents into a site-packages folder
3. zip the results into a `lambda_layer_payload.zip` file, clean up the non-zipped contents, and proceed to the next dir

Reply:
Is there any real need for the multiple layers? You already have a requirements.txt file in each module... if you structure your lambdas as packages instead of scripts, then terraform-aws-lambda will automatically package the dependencies for you...

See: https://github.com/claranet/terraform-aws-lambda/tree/master/tests/python-versions

Reply:
The big driver for layers was to support python-ldap since there is a bin (bins?) that needs to be compiled as part of the [package install](https://www.python-ldap.org/en/latest/installing.html#build-prerequisites). I could have limited the layer generation logic to just the lambdas that had python-lambda as a req but figured it was just easier to not discriminate

Reply:
Right, I'm just saying we only need _one_ layer with the openldap-devel package. That layer can be used by every lambda that needs it. The python bindings can be in the lambda package rather than the layer. No need for a layer per-lambda...

Reply:
I went ahead and split out the layer creation logic into separate files and [wrapped it in terraform](https://github.com/plus3it/terraform-aws-ldap-maintainer/tree/init/modules/create_layer). I tried to make this module flexible enough to potentially be used in other projects that require lambda layers